src/qforte/utils/qubit_tapering.py

```diff
@@ -198,7 +202,91 @@ def find_parity_check_matrix_kernel(n_qubits, n_strings, prt_chck_mtrx):
+
+    Arguments:
+
+    basis: uint8 numpy array
+        The rows of this matrix constitute a basis of the kernel of the parity check matrix.
+
+    Returns:
+
+    commute: uint8 numpy array
+        commute[i,j] = 0 (1) means that the ith and jth basis vectors commute (anticommute).
+    """
+
+    ## Construct symplectic bilinear form matrix B, which acts on binary vectors.
+    ## v * B * w.T = 0/1 means v and w commute/anticommute.
+
+    zero_matrix = np.zeros((int(basis.shape[1]/2), int(basis.shape[1]/2)), dtype=np.uint8)
+    identity_matrix = np.identity(int(basis.shape[1]/2), dtype=np.uint8)
+    blnr_frm_mtrx = np.block([
+        [zero_matrix    , identity_matrix],
+        [identity_matrix, zero_matrix    ]
+        ])
+
+    ## For all pairs of ker(E) basis vectors, compute commute[v, w] = v * B * w.T
+
+    commute = np.matmul(basis, np.matmul(blnr_frm_mtrx, basis.T))
+    commute %= 2
+
+    return commute
+
+def Symplectic_Gram_Schmidt_Orthogonalization(basis, commute):
+    """
+    This function performs a symplectic Gram-Schmidt orthogonalization on a given
+    set of binary basis vectors.
+
+    Arguments:
+
+    basis: uint8 numpy array
+        The rows of this matrix constitute a basis of the kernel of the parity check matrix.
+
+    commute: uint8 numpy array
+        commute[i,j] = 0 (1) means that the ith and jth basis vectors commute (anticommute).
+
+    Returns:
+
+    SGSO_basis: uint8 numpy array
+        The rows of this matrix constitute the Gram-Schmidt orthogonalized basis of the
+        kernel of the parity check matrix.
+
+    SGSO_commute: uint8 numpy array
+        SGSO_commute[i,j] = 0 (1) means that the ith and jth Gram-Schmidt orthogonalized basis
+        vectors commute (anticommute).
+
+    """
+
+    SGSO_basis = basis.copy()
+    SGSO_commute = commute.copy()
+
+    processed = []
+    for pauli_1 in range(SGSO_basis.shape[0]):
+        if pauli_1 not in processed:
+            processed.append(pauli_1)
+            anticommute = np.argwhere(SGSO_commute[pauli_1,:])
+            if anticommute.shape[0] != 0:
+                for i in anticommute:
+                    pauli_2 = int(i)
+                    if pauli_2 not in processed:
+                        processed.append(pauli_2)
+                        for pauli in range(SGSO_basis.shape[0]):
+                            if pauli not in processed:
+                                if SGSO_commute[pauli, pauli_1] and SGSO_commute[pauli, pauli_2]:
+                                    SGSO_basis[pauli] ^= SGSO_basis[pauli_1] ^ SGSO_basis[pauli_2]
+                                elif SGSO_commute[pauli, pauli_1] and not SGSO_commute[pauli, pauli_2]:
+                                    SGSO_basis[pauli] ^= SGSO_basis[pauli_2]
+                                elif not SGSO_commute[pauli, pauli_1] and SGSO_commute[pauli, pauli_2]:
+                                    SGSO_basis[pauli] ^= SGSO_basis[pauli_1]
```

Comment:
It looks like you can rewrite this as
```
if SGSO_commute[pauli, pauli_2]:
    SGSO_basis[pauli] ^= SGSO_basis[pauli_1]
if SGSO_commute[pauli, pauli_1]:
    SGSO_basis[pauli] ^= SGSO_basis[pauli_2]
```
This seems simpler to me.

Reply:
This is indeed very compact!